oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,305 @@
+
+We expect the following packages - which are in use in the vast majority of MFEs today - could all be shared between MFEs.
+
+.. list-table::
+   :widths: 30 20 50
+
+   * - **Package**
+     - **Estimated Size**
+     - **Notes**
+   * - @edx/brand
+     - Variable
+     -
+   * - @edx/frontend-component-footer
+     - 88.1k
+     -
+   * - @edx/frontend-component-header
+     - 156.9k
+     -
+   * - @edx/frontend-platform
+     - 355.3k
+     -
+   * - @openedx/paragon (or @edx/paragon)
+     - Big
+     - bundlephobia.com choked on the paragon build, but we expect it’s the largest thing here.  That said, today, Paragon components are dropped during treeshaking
+   * - classnames
+     - 0.8k
+     -
+   * - core-js
+     - 241.1k
+     -
+   * - prop-types
+     - 0.9k
+     -
+   * - react
+     - 6.4k
+     -
+   * - react-dom
+     - 130.2k
+     -
+   * - react-redux
+     - 11.2k
+     -
+   * - react-router
+     - 58.9k
+     -
+   * - react-router-dom
+     - 77.1k
+     -
+   * - redux
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit.  This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~892.2k unzipped
+
+Sharing these dependencies has several important positive effects:
+
+- Enforces consistency across all types of dependencies
+- Reduces runtime bundle size
+
+Note that "build time" and "dependency maintenance" are not mentioned above.
+
+Interestingly, because of the way webpack module federation's dependency sharing is implemented, this may not have a large impact on build time or dependency maintenance, as "guest" MFEs still need to have the dependencies installed for the build to succeed, and artifacts for them will still be created - they just won't be used at runtime.  Using webpack externals instead of module federation ensures that the dependencies won't be bundled/part of the build, but they still need to be dev dependencies in the repository for IDE hinting, auto completion, linting, etc.
+
+Enforcing Dependency Consistency
+================================
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity.
+
+We expect that this may need to take a number of possible forms:
+
+- Process to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent.
+- Process to migrate through breaking changes in third-party dependencies.
+- Best practices to ensure we minimize breaking changes in our own libraries (such as Paragon).
```

Comment:
We should already be doing that, but after this there'll be a more pressing reason for it, indeed.

Reply:
I've added some specific recommendations around this.